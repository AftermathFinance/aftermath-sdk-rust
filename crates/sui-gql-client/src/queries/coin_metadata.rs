use super::Error;
use crate::{GraphQlClient, GraphQlResponseExt, schema};

pub async fn query<C: GraphQlClient>(
    client: &C,
    type_: impl AsRef<str>,
) -> Result<(Option<u8>, Option<String>, Option<String>), Error<C::Error>> {
    let data = client
        .query::<Query, _>(QueryVariables {
            coin_type: type_.as_ref(),
        })
        .await
        .map_err(Error::Client)?
        .try_into_data()?;
    graphql_extract::extract!(data => {
        coin_metadata? {
            decimals
            name
            symbol
        }
    });
    Ok((decimals.map(|d| d as u8), name, symbol))
}

#[cfg(test)]
#[allow(clippy::unwrap_used)]
#[test]
fn gql_output() {
    use cynic::QueryBuilder as _;

    let vars = QueryVariables {
        coin_type: "0x2::sui::Sui",
    };
    let operation = Query::build(vars);
    insta::assert_snapshot!(operation.query, @r###"
    query Query($coinType: String!) {
      coinMetadata(coinType: $coinType) {
        decimals
        name
        symbol
      }
    }
    "###);
}

#[derive(cynic::QueryVariables, Debug)]
struct QueryVariables<'a> {
    coin_type: &'a str,
}

#[derive(cynic::QueryFragment, Debug)]
#[cynic(graphql_type = "Query", variables = "QueryVariables")]
struct Query {
    #[arguments(coinType: $coin_type)]
    coin_metadata: Option<CoinMetadata>,
}

#[derive(cynic::QueryFragment, Debug)]
struct CoinMetadata {
    decimals: Option<i32>,
    name: Option<String>,
    symbol: Option<String>,
}
