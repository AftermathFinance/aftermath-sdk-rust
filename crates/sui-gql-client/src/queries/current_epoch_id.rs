use af_sui_types::Version;
use cynic::{GraphQlResponse, QueryFragment};

use super::Error;
use crate::{missing_data, schema, GraphQlClient, GraphQlResponseExt as _};

pub async fn query<C: GraphQlClient>(client: &C) -> Result<u64, Error<C::Error>> {
    let curr: GraphQlResponse<Query> = client.query(()).await.map_err(Error::Client)?;
    let curr_epoch_id = curr
        .try_into_data()?
        .ok_or_else(|| missing_data!("No data"))?
        .epoch
        .ok_or_else(|| missing_data!("epoch"))?
        .epoch_id;
    Ok(curr_epoch_id)
}

#[cfg(test)]
#[allow(clippy::unwrap_used)]
#[test]
fn init_gql_output() {
    use cynic::QueryBuilder as _;
    let operation = Query::build(());
    insta::assert_snapshot!(operation.query);
}

#[derive(QueryFragment, Clone, Debug)]
#[cynic(graphql_type = "Query")]
struct Query {
    epoch: Option<Epoch>,
}

#[derive(QueryFragment, Clone, Debug)]
struct Epoch {
    epoch_id: Version,
}
