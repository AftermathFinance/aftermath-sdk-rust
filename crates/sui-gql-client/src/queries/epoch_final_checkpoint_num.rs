use cynic::{GraphQlResponse, QueryFragment};

use super::Error;
use crate::scalars::UInt53;
use crate::{missing_data, schema, GraphQlClient, GraphQlResponseExt as _};

pub async fn query<C: GraphQlClient>(client: &C, epoch_id: u64) -> Result<u64, Error<C::Error>> {
    let result: GraphQlResponse<Query> = client
        .query(Variables {
            id: Some(epoch_id.into()),
        })
        .await
        .map_err(Error::Client)?;
    Ok(result
        .try_into_data()?
        .ok_or_else(|| missing_data!("No data"))?
        .epoch
        .ok_or_else(|| missing_data!("epoch"))?
        .checkpoints
        .nodes
        .pop()
        .ok_or_else(|| missing_data!("checkpoints"))?
        .sequence_number
        .into())
}

#[cfg(test)]
#[allow(clippy::unwrap_used)]
#[test]
fn init_gql_output() {
    use cynic::QueryBuilder as _;
    let operation = Query::build(Variables { id: None });
    insta::assert_snapshot!(operation.query);
}

#[derive(cynic::QueryVariables, Clone, Debug)]
struct Variables {
    id: Option<UInt53>,
}

#[derive(QueryFragment, Clone, Debug)]
#[cynic(graphql_type = "Query", variables = "Variables")]
struct Query {
    #[arguments(id: $id)]
    epoch: Option<Epoch>,
}

#[derive(QueryFragment, Clone, Debug)]
struct Epoch {
    #[arguments(last: 1)]
    checkpoints: CheckpointConnection,
}

#[derive(QueryFragment, Clone, Debug)]
struct CheckpointConnection {
    nodes: Vec<Checkpoint>,
}

#[derive(QueryFragment, Clone, Debug)]
struct Checkpoint {
    sequence_number: UInt53,
}
