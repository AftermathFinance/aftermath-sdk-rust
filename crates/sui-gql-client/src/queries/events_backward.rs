use af_sui_types::Address as SuiAddress;
use sui_gql_schema::scalars;

use super::fragments::MoveValueRaw;
use super::Error;
use crate::{schema, GraphQlClient, GraphQlResponseExt as _};

#[derive(cynic::InputObject, Debug, Clone)]
pub struct EventFilter {
    pub sender: Option<SuiAddress>,
    pub transaction_digest: Option<String>,
    pub emitting_module: Option<String>,
    pub event_type: Option<String>,
}

#[derive(cynic::QueryFragment, Debug, Clone)]
pub struct EventEdge {
    pub node: Event,
    pub cursor: String,
}

#[derive(cynic::QueryFragment, Debug, Clone)]
pub struct Event {
    pub timestamp: Option<scalars::DateTime>,
    pub contents: Option<MoveValueRaw>,
}

/// Return a single page of events + cursors and a flag indicating if there's a previous page.
///
/// If `page_size` is left `None`, the server decides the size of the page.
///
/// The edges are returned in reverse order of which they where returned by the server
pub async fn query<C: GraphQlClient>(
    client: &C,
    filter: Option<EventFilter>,
    cursor: Option<String>,
    page_size: Option<u32>,
) -> super::Result<(Vec<EventEdge>, bool), C> {
    let vars = Variables {
        filter,
        before: cursor,
        last: page_size.map(|v| v as i32),
    };
    let data: Option<Query> = client
        .query(vars)
        .await
        .map_err(Error::Client)?
        .try_into_data()?;
    graphql_extract::extract!(data => {
        events {
            edges
            page_info {
                has_previous_page
            }
        }
    });
    Ok((edges, has_previous_page))
}

// =============================================================================
//  Initial query
// =============================================================================

#[derive(cynic::QueryVariables, Debug, Clone)]
struct Variables {
    last: Option<i32>,
    before: Option<String>,
    filter: Option<EventFilter>,
}

#[derive(cynic::QueryFragment, Debug)]
#[cynic(graphql_type = "Query", variables = "Variables")]
struct Query {
    #[arguments(before: $before, filter: $filter, last: $last)]
    events: EventConnection,
}

#[cfg(test)]
#[allow(clippy::unwrap_used)]
#[test]
fn init_gql_output() {
    use cynic::QueryBuilder as _;
    let filter = EventFilter {
        sender: None,
        transaction_digest: None,
        emitting_module: None,
        event_type: None,
    };
    let vars = Variables {
        filter: Some(filter),
        before: None,
        last: None,
    };
    let operation = Query::build(vars);
    insta::assert_snapshot!(operation.query, @r###"
    query Query($last: Int, $before: String, $filter: EventFilter) {
      events(before: $before, filter: $filter, last: $last) {
        edges {
          node {
            timestamp
            contents {
              type {
                repr
              }
              bcs
            }
          }
          cursor
        }
        pageInfo {
          hasPreviousPage
        }
      }
    }
    "###);
}

// =============================================================================
//  Inner query fragments
// =============================================================================

#[derive(cynic::QueryFragment, Clone, Debug)]
struct EventConnection {
    edges: Vec<EventEdge>,
    page_info: HasPreviousPage,
}

#[derive(cynic::QueryFragment, Clone, Debug)]
#[cynic(graphql_type = "PageInfo")]
pub struct HasPreviousPage {
    pub has_previous_page: bool,
}
