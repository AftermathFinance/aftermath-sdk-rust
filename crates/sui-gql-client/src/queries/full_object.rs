use af_sui_types::sui::object::Object;
use af_sui_types::ObjectId;
use cynic::{QueryFragment, QueryVariables};

use super::Error;
use crate::{missing_data, scalars, schema, GraphQlClient, GraphQlResponseExt as _};

/// Get the full [`Object`] contents from the server at a certain version or the latest if not
/// specified.
pub async fn query<C>(
    client: &C,
    object_id: ObjectId,
    version: Option<u64>,
) -> Result<Object, Error<C::Error>>
where
    C: GraphQlClient,
{
    let data = client
        .query::<Query, _>(Variables {
            address: object_id,
            version: version.map(From::from),
        })
        .await
        .map_err(Error::Client)?
        .try_into_data()?
        .ok_or(missing_data!("Null data in response"))?;

    let object = data
        .object
        .ok_or(missing_data!("Object not found"))?
        .bcs
        .ok_or(missing_data!("No object BCS"))?;
    Ok(object.into_inner())
}

#[derive(QueryVariables, Clone, Debug)]
struct Variables {
    address: ObjectId,
    version: Option<af_sui_types::Version>,
}

#[derive(QueryFragment, Clone, Debug)]
#[cynic(variables = "Variables")]
struct Query {
    #[arguments(address: $address, version: $version)]
    object: Option<GqlObject>,
}

#[derive(QueryFragment, Clone, Debug)]
#[cynic(graphql_type = "Object")]
struct GqlObject {
    bcs: Option<scalars::Base64Bcs<Object>>,
}

#[cfg(test)]
#[allow(clippy::unwrap_used)]
#[test]
fn gql_output() {
    use cynic::QueryBuilder as _;

    let vars = Variables {
        address: ObjectId::new(rand::random()),
        version: None,
    };
    let operation = Query::build(vars);
    insta::assert_snapshot!(operation.query);
}
