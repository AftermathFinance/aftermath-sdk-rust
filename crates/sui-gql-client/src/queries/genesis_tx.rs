use af_sui_types::TransactionData;
use cynic::GraphQlResponse;

use super::Error;
use crate::{missing_data, scalars, schema, GraphQlClient, GraphQlResponseExt as _};

pub async fn query<C: GraphQlClient>(client: &C) -> Result<TransactionData, Error<C::Error>> {
    let result: GraphQlResponse<Query> = client
        .query(Variables { id: Some(0) })
        .await
        .map_err(Error::Client)?;
    let wrapped = result
        .try_into_data()?
        .ok_or_else(|| missing_data!("No data"))?
        .epoch
        .ok_or_else(|| missing_data!("epoch"))?
        .transaction_blocks
        .nodes
        .pop()
        .ok_or_else(|| missing_data!("transaction_blocks"))?
        .bcs
        .ok_or_else(|| missing_data!("bcs"))?;
    Ok(wrapped.into_inner())
}

#[derive(cynic::QueryVariables, Clone, Debug)]
struct Variables {
    id: Option<af_sui_types::Version>,
}

#[derive(cynic::QueryFragment, Clone, Debug)]
#[cynic(graphql_type = "Query", variables = "Variables")]
struct Query {
    #[arguments(id: $id)]
    epoch: Option<Epoch>,
}

#[derive(cynic::QueryFragment, Clone, Debug)]
struct Epoch {
    #[arguments(first: 1)]
    transaction_blocks: TransactionBlockConnection,
}

#[derive(cynic::QueryFragment, Clone, Debug)]
struct TransactionBlockConnection {
    nodes: Vec<TransactionBlock>,
}

#[derive(cynic::QueryFragment, Clone, Debug)]
struct TransactionBlock {
    bcs: Option<scalars::Base64Bcs<TransactionData>>,
}

#[cfg(test)]
#[allow(clippy::unwrap_used)]
#[test]
fn init_gql_output() {
    use cynic::QueryBuilder as _;
    let operation = Query::build(Variables { id: None });
    insta::assert_snapshot!(operation.query);
}
