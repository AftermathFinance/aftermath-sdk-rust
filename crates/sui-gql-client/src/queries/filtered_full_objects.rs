use std::collections::HashMap;

use af_sui_types::{Address, Object, ObjectId, TypeTag};

use super::fragments::{ObjectFilterV2, PageInfoForward};
use crate::queries::Error;
use crate::{missing_data, scalars, schema, GraphQlClient, Paged};

pub(super) async fn query<C: GraphQlClient>(
    client: &C,
    owner: Option<Address>,
    type_: Option<TypeTag>,
    page_size: Option<u32>,
) -> Result<HashMap<ObjectId, Object>, Error<C::Error>> {
    let filter = ObjectFilterV2 {
        owner,
        type_: type_.map(scalars::TypeTag),
        ..Default::default()
    };
    let vars = Variables {
        after: None,
        first: page_size.map(|v| v.try_into().unwrap_or(i32::MAX)),
        filter: Some(filter),
    };

    let (init, pages) = client
        .query_paged::<Query>(vars)
        .await
        .map_err(Error::Client)?
        .try_into_data()?
        .ok_or(missing_data!("Empty response data"))?;

    let mut raw_objs = HashMap::new();
    let init_nodes = init.objects.nodes;
    let page_nodes = pages.into_iter().flat_map(|q| q.objects.nodes);
    for ObjectGql { id, object } in init_nodes.into_iter().chain(page_nodes) {
        let wrapped = object.ok_or(missing_data!("Bcs for object {id}"))?;
        raw_objs.insert(id, wrapped.into_inner());
    }

    Ok(raw_objs)
}

#[derive(cynic::QueryVariables, Clone, Debug)]
struct Variables {
    filter: Option<ObjectFilterV2>,
    after: Option<String>,
    first: Option<i32>,
}

#[derive(cynic::QueryFragment, Clone, Debug)]
#[cynic(variables = "Variables")]
struct Query {
    #[arguments(filter: $filter, first: $first, after: $after)]
    objects: ObjectConnection,
}

impl Paged for Query {
    type Input = Variables;

    type NextInput = Variables;

    type NextPage = Self;

    fn next_variables(&self, mut prev_vars: Self::Input) -> Option<Self::NextInput> {
        let PageInfoForward {
            has_next_page,
            end_cursor,
        } = &self.objects.page_info;
        if *has_next_page {
            prev_vars.after.clone_from(end_cursor);
            Some(prev_vars)
        } else {
            None
        }
    }
}

// =============================================================================
//  Inner query fragments
// =============================================================================

/// `ObjectConnection` where the `Object` fragment does take any parameters.
#[derive(cynic::QueryFragment, Clone, Debug)]
struct ObjectConnection {
    nodes: Vec<ObjectGql>,
    page_info: PageInfoForward,
}

#[derive(cynic::QueryFragment, Debug, Clone)]
#[cynic(graphql_type = "Object")]
struct ObjectGql {
    #[cynic(rename = "address")]
    id: ObjectId,
    #[cynic(rename = "bcs")]
    object: Option<scalars::Base64Bcs<Object>>,
}

#[cfg(test)]
#[allow(clippy::unwrap_used)]
#[test]
fn gql_output() -> color_eyre::Result<()> {
    use cynic::QueryBuilder as _;

    let vars = Variables {
        filter: None,
        after: None,
        first: None,
    };

    let operation = Query::build(vars);
    insta::assert_snapshot!(operation.query, @r###"
    query Query($filter: ObjectFilter, $after: String, $first: Int) {
      objects(filter: $filter, first: $first, after: $after) {
        nodes {
          address
          bcs
        }
        pageInfo {
          hasNextPage
          endCursor
        }
      }
    }
    "###);
    Ok(())
}
