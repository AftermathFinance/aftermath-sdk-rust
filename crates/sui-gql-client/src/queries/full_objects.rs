use std::collections::HashMap;

use af_sui_types::{Object, ObjectId};
use futures::{StreamExt as _, TryStreamExt as _};
use graphql_extract::extract;
use itertools::{Either, Itertools as _};
use sui_gql_schema::scalars::Base64Bcs;

use super::fragments::{ObjectFilter, ObjectKey, PageInfo, PageInfoForward};
use super::stream;
use crate::queries::Error;
use crate::{missing_data, schema, GraphQlClient, GraphQlResponseExt as _};

pub(super) async fn query<C: GraphQlClient>(
    client: &C,
    objects: impl IntoIterator<Item = (ObjectId, Option<u64>)> + Send,
    page_size: Option<u32>,
) -> Result<HashMap<ObjectId, Object>, Error<C::Error>> {
    // To keep track of all ids requested.
    let mut requested = vec![];

    let (object_ids, object_keys) = objects
        .into_iter()
        .inspect(|(id, _)| requested.push(*id))
        .partition_map(|(id, v)| {
            v.map_or(Either::Left(id), |n| {
                Either::Right(ObjectKey {
                    object_id: id,
                    version: n,
                })
            })
        });

    #[expect(
        deprecated,
        reason = "TODO: build query from scratch with new ObjectFilter and Query.multiGetObjects"
    )]
    let filter = ObjectFilter {
        object_ids: Some(object_ids),
        object_keys: Some(object_keys),
        ..Default::default()
    };
    let vars = Variables {
        after: None,
        first: page_size.map(|v| v.try_into().unwrap_or(i32::MAX)),
        filter: Some(filter),
    };

    let raw_objs: HashMap<_, _> = stream::forward(client, vars, request)
        .map(|r| -> super::Result<_, C> {
            let (id, obj) = r?;
            Ok((id, obj.ok_or_else(|| missing_data!("BCS for {id}"))?))
        })
        .try_collect()
        .await?;

    // Ensure all requested objects were returned
    for id in requested {
        raw_objs
            .contains_key(&id)
            .then_some(())
            .ok_or(missing_data!("Object {id}"))?;
    }

    Ok(raw_objs)
}

async fn request<C: GraphQlClient>(
    client: &C,
    vars: Variables,
) -> super::Result<
    stream::Page<impl Iterator<Item = super::Result<(ObjectId, Option<Object>), C>> + 'static>,
    C,
> {
    let data = client
        .query::<Query, _>(vars)
        .await
        .map_err(Error::Client)?
        .try_into_data()?;

    extract!(data => {
        objects {
            nodes[] {
                id
                object
            }
            page_info
        }
    });
    Ok(stream::Page::new(
        page_info,
        nodes.map(|r| -> super::Result<_, C> {
            let (id, obj) = r?;
            Ok((id, obj.map(Base64Bcs::into_inner)))
        }),
    ))
}

#[derive(cynic::QueryVariables, Clone, Debug)]
struct Variables {
    filter: Option<ObjectFilter>,
    after: Option<String>,
    first: Option<i32>,
}

impl stream::UpdatePageInfo for Variables {
    fn update_page_info(&mut self, info: &PageInfo) {
        self.after.clone_from(&info.end_cursor);
    }
}

#[derive(cynic::QueryFragment, Clone, Debug)]
#[cynic(variables = "Variables")]
struct Query {
    #[arguments(filter: $filter, first: $first, after: $after)]
    objects: ObjectConnection,
}

// =============================================================================
//  Inner query fragments
// =============================================================================

/// `ObjectConnection` where the `Object` fragment does take any parameters.
#[derive(cynic::QueryFragment, Clone, Debug)]
struct ObjectConnection {
    nodes: Vec<ObjectGql>,
    page_info: PageInfoForward,
}

#[derive(cynic::QueryFragment, Debug, Clone)]
#[cynic(graphql_type = "Object")]
struct ObjectGql {
    #[cynic(rename = "address")]
    id: ObjectId,
    #[cynic(rename = "bcs")]
    object: Option<Base64Bcs<Object>>,
}

#[cfg(test)]
#[allow(clippy::unwrap_used)]
#[test]
fn gql_output() -> color_eyre::Result<()> {
    use cynic::QueryBuilder as _;

    let vars = Variables {
        filter: Some(ObjectFilter {
            object_ids: Some(vec![
                "0x4264c07a42f9d002c1244e43a1f0fa21c49e4a25c7202c597b8476ef6bb57113".parse()?,
                "0x60d1a85f81172a7418206f4b16e1e07e40c91cf58783f63f18a25efc81442dcb".parse()?,
            ]),
            ..Default::default()
        }),
        after: None,
        first: None,
    };

    let operation = Query::build(vars);
    insta::assert_snapshot!(operation.query, @r###"
    query Query($filter: ObjectFilter, $after: String, $first: Int) {
      objects(filter: $filter, first: $first, after: $after) {
        nodes {
          address
          bcs
        }
        pageInfo {
          hasNextPage
          endCursor
        }
      }
    }
    "###);
    Ok(())
}
