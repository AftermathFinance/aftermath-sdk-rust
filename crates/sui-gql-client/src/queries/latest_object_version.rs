use af_sui_types::ObjectId;

use crate::queries::Error;
use crate::{missing_data, scalars, schema, GraphQlClient, GraphQlResponseExt as _};

pub async fn query<C>(client: &C, object_id: ObjectId) -> Result<(u64, u64), Error<C::Error>>
where
    C: GraphQlClient,
{
    let data = client
        .query::<Query, _>(Variables { object_id })
        .await
        .map_err(Error::Client)?
        .try_into_data()?
        .ok_or(missing_data!("Null data in response"))?;

    Ok((
        data.checkpoint
            .ok_or(missing_data!("Checkpoint"))?
            .sequence_number
            .0,
        data.object
            .ok_or(missing_data!("Object not found"))?
            .version
            .0,
    ))
}

#[derive(cynic::QueryVariables, Debug)]
struct Variables {
    object_id: ObjectId,
}

#[derive(cynic::QueryFragment, Debug)]
#[cynic(variables = "Variables")]
struct Query {
    checkpoint: Option<Checkpoint>,

    #[arguments(address: $object_id)]
    object: Option<Object>,
}

#[derive(cynic::QueryFragment, Debug)]
struct Object {
    version: scalars::UInt53,
}

#[derive(cynic::QueryFragment, Debug)]
struct Checkpoint {
    sequence_number: scalars::UInt53,
}

#[cfg(test)]
#[allow(clippy::unwrap_used)]
#[test]
fn gql_output() {
    use cynic::QueryBuilder as _;

    let vars = Variables {
        object_id: ObjectId::new(rand::random()),
    };
    let operation = Query::build(vars);
    insta::assert_snapshot!(operation.query);
}
