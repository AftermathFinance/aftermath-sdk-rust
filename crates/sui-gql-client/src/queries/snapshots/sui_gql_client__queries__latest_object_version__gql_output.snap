---
source: crates/sui-gql-client/src/queries/latest_object_version.rs
expression: operation.query
---
query Query($objectId: SuiAddress!) {
  checkpoint {
    sequenceNumber
  }
  object(address: $objectId) {
    version
  }
}
